src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/PropertyRef.cs

```diff
@@ -1,21 +1,113 @@
+        /// A custom hashcode produced from the UTF-8 encoded property name.
+        /// </summary>
+        public readonly ulong Key = key;
+
+        /// <summary>
+        /// The <see cref="JsonPropertyInfo"/> associated with the property name, if available.
+        /// </summary>
+        public readonly JsonPropertyInfo? Info = info;
+
+        /// <summary>
+        /// Caches a heap allocated copy of the UTF-8 encoded property name.
+        /// </summary>
+        public readonly byte[] Utf8PropertyName = utf8PropertyName;
+
+        public bool Equals(PropertyRef other) => Equals(other.Utf8PropertyName, other.Key);
+        public override bool Equals(object? obj) => obj is PropertyRef other && Equals(other);
+        public override int GetHashCode() => Key.GetHashCode();
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public bool Equals(ReadOnlySpan<byte> propertyName, ulong key)
         {
-            Key = key;
-            Info = info;
-            NameFromJson = nameFromJson;
+            if (key == Key)
+            {
+                // We compare the whole name, although we could skip the first 7 bytes (but it's not any faster)
+                if (propertyName.Length <= PropertyNameKeyLength ||
+                    propertyName.SequenceEqual(Utf8PropertyName))
+                {
+                    return true;
+                }
+            }
+
+            return false;
         }
 
-        public readonly ulong Key;
-        public readonly JsonPropertyInfo Info;
+        /// <summary>
+        /// Get a key from the property name.
+        /// The key consists of the first 7 bytes of the property name and then the length.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static ulong GetKey(ReadOnlySpan<byte> name)
+        {
+            ulong key;
+
+            ref byte reference = ref MemoryMarshal.GetReference(name);
+            int length = name.Length;
 
-        // NameFromJson may be different than Info.NameAsUtf8Bytes when case insensitive is enabled.
-        public readonly byte[] NameFromJson;
+            if (length > 7)
+            {
+                key = Unsafe.ReadUnaligned<ulong>(ref reference) & 0x00ffffffffffffffL;
+                key |= (ulong)Math.Min(length, 0xff) << 56;
+            }
+            else
+            {
+                key =
+                    length > 5 ? Unsafe.ReadUnaligned<uint>(ref reference) | (ulong)Unsafe.ReadUnaligned<ushort>(ref Unsafe.Add(ref reference, 4)) << 32 :
+                    length > 3 ? Unsafe.ReadUnaligned<uint>(ref reference) :
+                    length > 1 ? Unsafe.ReadUnaligned<ushort>(ref reference) : 0UL;
+                key |= (ulong)length << 56;
+
+                if ((length & 1) != 0)
+                {
+                    int offset = length - 1;
+                    key |= (ulong)Unsafe.Add(ref reference, offset) << (offset * 8);
+                }
```

Comment:
If you're really trying to save cycles here, you could experiment with a switch something like:
```C#
key = (ulong)length << 56;
switch (length)
{
    case 0: break;
    case 1: goto OddLength;
    case 2: key |= Unsafe.ReadUnaligned<ushort>(ref reference); goto ComputedKey;
    case 3: key |= Unsafe.ReadUnaligned<ushort>(ref reference); goto OddLength;
    case 4: key |= Unsafe.ReadUnaligned<uint>(ref reference); goto ComputedKey;
    case 5: key |= Unsafe.ReadUnaligned<uint>(ref reference); goto OddLength;
    case 6: key |= Unsafe.ReadUnaligned<uint>(ref reference) | (ulong)Unsafe.ReadUnaligned<ushort>(ref Unsafe.Add(ref reference, 4)) << 32; goto ComputedKey;
    case 7: key |= Unsafe.ReadUnaligned<uint>(ref reference) | (ulong)Unsafe.ReadUnaligned<ushort>(ref Unsafe.Add(ref reference, 4)) << 32; goto OddLength;
    case _: key = (Unsafe.ReadUnaligned<ulong>(ref reference) & 0x00ffffffffffffffL) | ((ulong)Math.Min(length, 0xff) << 56); goto ComputedKey;
}

OddLength:
int offset = length - 1;
key |= (ulong)Unsafe.Add(ref reference, offset) << (offset * 8);

ComputedKey:;
```

Reply:
The particular algorithm is being copied as-is from JsonTypeInfo. I experimented with a few things, including using the HashCode type and Marvin but didn't see any noticeable differences in performance.

Reply:
I've incorporated your suggestion and then tweaked the algorithm somewhat to simplify length embedding. It now always stores the least significant byte of the length.